[Docs][zh]Localize contribution_guidelines/testing (#6307)
kubernetes-sigs/kueue: site/content/zh-CN/docs/contribution_guidelines/testing.md

---
title: "运行和调试测试"
linkTitle: "运行测试"
weight: 25
description: >
  运行和调试测试
---

> ## 运行预提交验证测试

Reviewer: ```suggestion
## 运行预提交验证测试 {#running-presubmission-verification-tests}
```
Add anchors to all headings 2 and 3

```shell
make verify
```

## 运行单元测试
运行所有单元测试：
```shell
make test
```

> 运行 webhooks 的单元测试：

Author: Thanks, updated.

```shell
go test ./pkg/webhooks
```

> 运行匹配 `TestValidateClusterQueue` 正则表达式的测试 [参考](https://pkg.go.dev/cmd/go#hdr-Testing_flags)：

Reviewer: ```suggestion
[参考文档](https://pkg.go.dev/cmd/go#hdr-Testing_flags)运行与 `TestValidateClusterQueue` 正则表达式匹配的测试：
```

```shell
go test ./pkg/webhooks -run TestValidateClusterQueue
```

### 使用竞态检测运行单元测试

使用 `-race` 启用 Go 内置的竞态检测器：
```shell
go test ./pkg/scheduler/preemption/ -race
```

### 使用压力测试运行单元测试

循环运行单元测试并收集失败情况：
```shell
# 安装 go stress
go install golang.org/x/tools/cmd/stress@latest
# 编译测试（可以添加 -race 进行竞态检测）
go test ./pkg/scheduler/preemption/ -c
# 循环运行并报告失败
stress ./preemption.test -test.run TestPreemption
```

> ## 运行集成测试

Reviewer: ```suggestion
## 运行集成测试 {#running-integration-tests}
```

```shell
make test-integration
```

> 关于运行测试子集，请参阅 [运行测试子集](#运行集成或-e2e-测试子集)。

Reviewer: ```suggestion
关于运行测试子集，请参阅[运行测试子集](#running-subset-of-integration-or-e2e-tests)。
```

Author: All have updated.

make test-multikueue-e2e
```

您可以通过设置 `E2E_K8S_FULL_VERSION` 变量来指定用于运行 e2e 测试的 Kubernetes 版本：
```shell
E2E_K8S_FULL_VERSION=1.33.1 make test-e2e
```

> 关于运行测试子集，请参阅 [运行测试子集](#运行集成或-e2e-测试子集)。

Reviewer: ```suggestion
关于运行测试子集，请参阅[运行测试子集](#running-subset-of-integration-or-e2e-tests)。
```
Do not translate any anchor.

## 增加日志详细程度
您可以使用 `TEST_LOG_LEVEL` 变量更改日志级别（例如，设置 -5 来增加详细程度）。
默认情况下，`TEST_LOG_LEVEL=-3`。

## 在 VSCode 中调试测试
可以在 VSCode 中调试单元测试和集成测试。
您需要安装 [Go 扩展](https://marketplace.visualstudio.com/items?itemName=golang.Go)。
现在您将在如下行上方看到 `run test | debug test` 文本按钮：
```go
func TestValidateClusterQueue(t *testing.T) {
```
您可以点击 `debug test` 来调试特定测试。

对于集成测试，需要额外的步骤。在 settings.json 中，您需要在 `go.testEnvVars` 内添加两个变量：
- 运行 `ENVTEST_K8S_VERSION=1.33 make envtest && ./bin/setup-envtest use $ENVTEST_K8S_VERSION -p path` 并将路径分配给 `KUBEBUILDER_ASSETS` 变量
- 将 `KUEUE_BIN` 设置为您的 Kueue 仓库克隆目录内的 `bin` 目录
```json
"go.testEnvVars": {
    "KUBEBUILDER_ASSETS": "<上面输出的路径>",
    "KUEUE_BIN": "<您的-kueue-文件夹路径>/bin",
  },
```

对于 e2e 测试，您也可以使用 [Ginkgo Test Explorer](https://marketplace.visualstudio.com/items?itemName=joselitofilho.ginkgotestexplorer)。您需要在 settings.json 中添加以下变量：
```json
 "ginkgotestexplorer.testEnvVars": {
        "KIND_CLUSTER_NAME": "kind",
        "WORKER1_KIND_CLUSTER_NAME": "kind-worker1",
        "MANAGER_KIND_CLUSTER_NAME": "kind-manager",
        "WORKER2_KIND_CLUSTER_NAME": "kind-worker2",

>         "KIND": "<您的_kueue_路径>/bin/kind",

Reviewer: ```suggestion
        "KIND": "<your_kueue_path>/bin/kind",
```
Not necessary to translate here.

    },
```

> 然后您可以使用 Ginkgo Test Explorer 的 GUI 来运行单个测试，前提是您已经启动了 kind 集群（有关说明，请参阅[这里](#将-e2e-测试附加到现有的-kind-集群)）。

Reviewer: do not translate any anchor.

## 将 e2e 测试附加到现有的 kind 集群
您可以使用以下方法来启动 kind 集群，然后从命令行或 VSCode 运行 e2e 测试，
将它们附加到现有集群。例如，假设您想测试一些 multikueue-e2e 测试。

运行 `E2E_RUN_ONLY_ENV=true make kind-image-build test-multikueue-e2e` 并等待 `Press Enter to cleanup.` 出现。

集群已准备就绪，现在您可以从另一个终端运行测试：
```shell

> <您的_kueue_路径>/bin/ginkgo --json-report ./ginkgo.report -focus "MultiKueue when Creating a multikueue admission check Should run a jobSet on worker if admitted" -r

Reviewer: ```suggestion
<your_kueue_path>/bin/ginkgo --json-report ./ginkgo.report -focus "MultiKueue when Creating a multikueue admission check Should run a jobSet on worker if admitted" -r
```

```
或从 VSCode 运行。

## 运行集成或 e2e 测试子集
### 使用 Ginkgo --focus 参数
```shell
GINKGO_ARGS="--focus=Scheduler" make test-integration
GINKGO_ARGS="--focus=Creating a Pod requesting TAS" make test-e2e
```
### 使用 ginkgo.FIt
如果您想专注于特定测试，可以将这些测试的
`ginkgo.It` 更改为 `ginkgo.FIt`。
有关更多详细信息，请参阅[这里](https://onsi.github.io/ginkgo/#focused-specs)。
然后其他测试将被跳过。

> 例如，您可以将

Reviewer: ```suggestion
例如，你可以将
```
Use 你 everywhere

make test-multikueue-e2e
```

> 您可以通过设置 `E2E_K8S_FULL_VERSION` 变量来指定用于运行 e2e 测试的 Kubernetes 版本：

Reviewer: ```suggestion
你可以通过设置 `E2E_K8S_FULL_VERSION` 变量来指定用于运行 e2e 测试的 Kubernetes 版本：
```
Search 您 and replace with 你 globally
